### DISCUSSION TITLE
get & set calibration_time with Lonsonho QS-Zigbee-C01 (and maybe others)

### DISCUSSION THREAD

**Discussion by okastl** [Koenkk/zigbee2mqtt#8781] [Devices]
Hi there

I have created an external converter for Lonsonho QS-Zigbee-C01 curtain/blinds controller, which allows to get/set calibration_time.

This makes calibration much easier, if you already know the runtime of your blinds! The manual calibration is quite time consuming, and I usually don't get it "right" on the first try.
IMHO it is much easier to time your blinds with a stop watch and enter the value with the html frontend or mqtt friendlyname/set {"calibration_time": S}.

Another side effect is the possibility to have a "poor man's tilt".
You can position to the desired height of the blinds, change calibration time to 1-20 seconds, position one step down, change calibration time back.

I attach the converter in case someone finds this useful. It should probably work with many other Tuya based curtain/blinds/shades controllers.

EDIT: New Version
[QS-Zigbee-C01-2.zip](https://github.com/Koenkk/zigbee2mqtt/files/13440673/QS-Zigbee-C01-2.zip)

**Comment by jjiik** [comment]
Sorry for my noob question - I'd love to avoid manual calibration routine. But I am really puzzled on how to use your snippet. Could you give offer a hint for me? Much appreciated...

**Reply by okastl** [reply]
[QS-Zigbee-C01-2.zip](https://github.com/Koenkk/zigbee2mqtt/files/13440673/QS-Zigbee-C01-2.zip)

I have attached the newest converter. Add it to external converters on your configuration.yaml:
```
external_converters:
  - QS-Zigbee-C01-2.js
```

**Comment by ArnonAfriat** [comment]
@okastl That is exactly what I'm looking for because the calibration is stuck, not really calibrating, and I was trying to edit the state.json file in the zigbee2mqtt folder, but it doesn't seem to be loaded anywhere so I'm not sure what it is used for...

Anyway, I'm not sure what I should do with the js file you've shared, where do I place it? which file should I edit and add the 'external_converters' property to, also, I'm not sure what to expect once it is configured correct;ly can you share a screenshot or something that indicates it worked ? is there a new functionality in the web UI? if so, where?

Thanks again

**Comment by okastl** [comment]
> Anyway, I'm not sure what I should do with the js file you've shared, where do I place it?

In your zigbee2mqtt data folder.

> which file should I edit and add the 'external_converters' property to

Add it to external converters on your configuration.yaml


> is there a new functionality in the web UI? if so, where?

Yes. See screenshot. If it does not work for you, you might have a different variant. In this case, you may need to add your variant to the external converter.

![image](https://github.com/Koenkk/zigbee2mqtt/assets/86207466/82262e8b-2950-45b3-8004-09b1be15b812)

**Reply by ArnonAfriat** [reply]
@okastl 
It was intuitive enough so I could figure it out myself, but your detailed explanation is spot on! Thank you

**Comment by YvoTuk** [comment]
This would be great if I could get it to work.
I've used your latest version (QS-Zigbee-C01-2.js) 
The file is here: /homeassistant/zigbee2mqtt/QS-Zigbee-C01-2.js

/homeassistant/zigbee2mqtt/configuration.yaml was modified, there was no external converters section, so I added:
external_converters:
  - QS-Zigbee-C01-2.js
 (tried it with and without the - )

Looking at the properties of the controller I'd say it is listed in your file:
![Rolluik](https://github.com/Koenkk/zigbee2mqtt/assets/78449057/ca752ec5-0b1f-420c-956d-8527bf521c19)

I've restarted HA several times, but there's no change, the calibration remains a number.

Am I overlooking something?

**Reply by okastl** [reply]
No, it should work. Which z2m version are you using?
I am not sure about home assistant integration, as I don't use it.

**Reply by YvoTuk** [reply]
Zigbee2MQTT version 1.36.1-1

But... I found it!
After installing the code I've restarted HA, assuming Z2M would restart as well.. It doesn't.

I've just restarted Z2M manually (in integrations) and BAM! It's working.

So for Home Assistant users, it's good to know you have to restart the integration, not HA

**Comment by margielm** [comment]
Hey man! thank you for this extension. I cant believe that it is not part of standard zigbee2mqtt. without it calibrating shutters is just impossible :/

**Comment by okastl** [comment]
@Koenkk what would be the easiest and most painless way to add this functionality to [zigbee-herdsman-converters](https://github.com/Koenkk/zigbee-herdsman-converters) without breaking anything?

**Reply by Koenkk** [reply]
Could you make a pull request by [clicking here](https://github.com/Koenkk/zigbee-herdsman-converters/edit/master/src/devices/lonsonho.ts)?

**Comment by DangerDan83** [comment]
Thank you so much, this let me calibrate the QS-Zigbee-S10-C04.
Works perfectly now!

**Comment by wdonders** [comment]
I would love to get this working, but I have a slightly different module, the MOES MS-108ZR
( https://www.zigbee2mqtt.io/devices/MS-108ZR.html#moes-ms-108zr )

I've placed the js file in the zigbee2mqtt directory, stopped the Z2M service, added the external_converters lines to the Z2M configuration.yaml and started the service again.
The js file seems to be read when I look at the Z2M logs, but it seems to have no effect on the exposed settings of the MOES module. I still can't change the calibration_time (get message "INVALID_DATA_TYPE" when I try)

So I thought maybe the script does not include/see my module as candidate, so I looked at the "Zigbee Model" on the "About" page of the module in Zigbee2MQTT in HA.
Displayed there is "TS130F" and "_TZ3000_1dd0d5yi", so I've added this to the fingerprints in the js script:
{modelID: 'TS130F', manufacturerName: '_TZ3000_1dd0d5yi'}

However, still no luck after restarting Z2M, stil can't change the calibration_time :-(
Am I doing something wrong? I'm puzzled..

**Reply by andrei-lazarov** [reply]
There were some recent changes. Please remove the external converter, and try latest Z2M (dev branch / edge app)
https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html

**Reply by okastl** [reply]
[QS-Zigbee-C01-3.js](https://github.com/user-attachments/files/30750468/QS-Zigbee-C01-3.js)
Here is an updated version. Have fun.

**Reply by wdonders** [reply]
Thank you! Much appreciated!
